### PR TITLE
DE5667 - Reintroduce mongoidable caching

### PR DESCRIPTION
**Pull request by krimsonkla**
https://rally1.rallydev.com/#/18313249162d/iterationstatus?detail=%2Fdefect%2F603685922641

### CODE REVIEW

**Review by jaurand — APPROVED** [review]

File: app/models/mongoidable/policy_relation.rb [hunk 1]
```diff
@@ -2,30 +2,34 @@
...
-      @abilities.merge(merge_requirements)
+    def inherited_abilities
+      merged_policy_requirements
     end
 
-    def merge_requirements
-      result = Mongoidable::Abilities.new(mongoidable_identity, self)
-      return result unless policy
+    def merged_policy_requirements
+      merged_policy_requirements = Mongoidable::Abilities.new(mongoidable_identity, self)
```
Comment (RealNobody) [inline app/models/mongoidable/policy_relation.rb:17]: This is a comment on what may just be a personal foible.

I don't like creating a local variable anywhere in the code that is the same as a method name (even in that method.)

I understand why you would do it, but I find it confusing in the code to look at a single line and have to think about/wonder:  Is this a method call or a variable?

File: app/queries/mongoidable/policy_query.rb [hunk 2]
```diff
@@ -68,6 +68,9 @@ def find_id
 
     memoize :object_for_index,
             :object_for_update,
-            :find_params
+            :object_for_show,
+            :object_for_destroy,
+            :unsafe_params
+    :find_params
```
Comment (RealNobody) [inline app/queries/mongoidable/policy_query.rb:74]: typo?

It looks like you probably missed a comma?

File: lib/mongoidable/current_ability.rb [hunk 3]
```diff
@@ -8,18 +8,71 @@ module Mongoidable
...
+        case type
+          when :all
+            @renew_instance = true
+            @renew_inherited = true
+            @renew_ancestral = true
+          when :inherited
+            @renew_inherited = true
+          when :instance
+            @renew_instance = true
+        when :ancestral
```
Comment (RealNobody) [inline lib/mongoidable/current_ability.rb:54]: formatting of this one `when` being off is visually confusing.

File: lib/mongoidable/current_ability.rb [hunk 4]
```diff
@@ -8,18 +8,71 @@ module Mongoidable
...
+    def current_ability(parent = @parent_model || nil)
+      @parent_model ||= parent
+      if @abilities.blank? || @renew
+        @abilities = @abilities&.empty_clone || Mongoidable::Abilities.new(mongoidable_identity, @parent_model || self)
+        run_before_ability_callbacks(:inherited)
+        @abilities.merge(inherited_abilities(@renew_inherited))
+        run_after_ability_callbacks(:inherited)
+
+        run_before_ability_callbacks(:ancestral)
+        @abilities.merge(ancestral_abilities(self.changed? || @renew_ancestral))
```
Comment (RealNobody) [inline lib/mongoidable/current_ability.rb:23]: I'm not seeing where `ancestral_abilities` is defined to take a parameter.  I see definitions without a parameter, but none with one.

I could be missing it, but if so, I could use some help finding it.

File: lib/mongoidable/current_ability.rb [hunk 5]
```diff
@@ -8,18 +8,71 @@ module Mongoidable
...
+        run_after_ability_callbacks(:ancestral)
+
+        run_before_ability_callbacks(:instance)
+        @abilities.merge(own_abilities(@renew_instance))
```
Comment (RealNobody) [inline lib/mongoidable/current_ability.rb:27]: I'm not seeing where `own_abilities` is defined to take a parameter.  I see definitions without a parameter, but none with one.

I could be missing it, but if so, I could use some help finding it.

File: lib/mongoidable/instance_abilities.rb [hunk 6]
```diff
@@ -23,5 +25,7 @@ def own_abilities
       end
       own_abilities
     end
+
+    memoize :own_abilities
```
Comment (RealNobody) [inline lib/mongoidable/instance_abilities.rb:29]: `memoizing` `own_abilities` in this file seems appropriate, but at odds with how it is called [here](https://github.com/CardTapp/mongoidable/pull/137/files#diff-3dc574e58966ab4db7b92c6937ce4679f407718362049a2f88867d7760c28cafR27)

File: lib/mongoidable/document.rb [hunk 7]
```diff
@@ -8,22 +8,23 @@ module Document
     include Mongoidable::ClassAbilities
     include Mongoidable::InstanceAbilities
     include Mongoidable::CurrentAbility
-    include Mongoidable::RelationsDirtyTracking
+    # include Mongoidable::RelationsDirtyTracking
```
Comment (RealNobody) [inline lib/mongoidable/document.rb:11]: Do we want to keep this comment?

**Review by RealNobody — COMMENTED** [review]
I feel like I must be missing something because there are some things in this PR that I find very hard to follow/troubling.

File: lib/mongoidable/current_ability.rb (hunk 4 above)
Comment (krimsonkla) [reply lib/mongoidable/current_ability.rb:23]:
```
Every memoized function (which initially was not accepting any arguments) has a (reload) argument you can pass in to bypass and reset the memoization:

def some_method
  Time.now
end
memoize :some_method
```

File: lib/mongoidable/current_ability.rb (hunk 5 above)
Comment (krimsonkla) [reply lib/mongoidable/current_ability.rb:27]: https://github.com/CardTapp/mongoidable/pull/137#discussion_r694327505

File: lib/mongoidable/instance_abilities.rb (hunk 6 above)
Comment (krimsonkla) [reply lib/mongoidable/instance_abilities.rb:29]: https://github.com/CardTapp/mongoidable/pull/137#discussion_r694327505

**Review by RealNobody — APPROVED** [review]
